src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java

```diff
@@ -426,8 +396,24 @@ private static boolean bytesEqual(byte[] bytes, int byteLength, byte[] name, int
 
         @Specialization(guards = "inBounds(matchData, index)")
         protected Object begin(RubyMatchData matchData, int index,
+                @Cached ConditionProfile negativeBeginProfile,
+                @Cached ConditionProfile multiByteCharacterProfile,
+                @Cached RopeNodes.SingleByteOptimizableNode singleByteOptimizableNode,
                 @CachedLibrary(limit = "2") RubyStringLibrary strings) {
-            return MatchDataNodes.begin(matchData, index, strings.getRope(matchData.source));
+            // Taken from org.jruby.RubyMatchData.
+
+            final Rope matchDataSourceRope = strings.getRope(matchData.source);
```

Comment:
Let's get the rope after the early negative check, so we don't even read it if we don't need it (same for `end`).